### PR TITLE
refine evaluation

### PR DESCRIPTION
**Pull request by vimar-gu**
1. Add `autocast` options to non-species-classification tasks.
2. Add a `classification.py`, which contains both zero-shot and few-shot classification, so that the evaluation will only need a single feature extraction. The few-shot results will be slightly different from the paper due to randomness, but the gap is within the standard deviation.

### CODE REVIEW

**Comment by egrace479** [comment]
@vimar-gu, does this change any options or instructions as far as the [README eval section](https://github.com/Imageomics/bioclip-2/blob/experiment/README.md#evaluation) is concerned that we may want to add a note there or in the news section? I think if we have changes that alter results from the paper, we probably want to note that somewhere at the root level -- this could also be documentation added in a follow-up PR.

File: src/evaluation/awa2.py [hunk 1]
```diff
@@ -74,12 +76,13 @@ def evaluate(
...
     it = iter(dataloader)
     y_pred, y_true = [], []
+    auto_cast = get_autocast(args.precision)
     for b in range(total):
         features, labels, ids = next(it)
         features = features.to(args.device)
         labels = labels.numpy()
         ids = ids.numpy()
-        with torch.no_grad():
+        with torch.no_grad() and auto_cast():
```
Comment (NetZissou) [inline src/evaluation/awa2.py]:
```suggestion
        with torch.no_grad(), auto_cast():
```

File: src/evaluation/fishnet.py [hunk 2]
```diff
@@ -74,12 +76,13 @@ def evaluate(
...
     it = iter(dataloader)
     y_pred, y_true = [], []
+    auto_cast = get_autocast(args.precision)
     for b in range(total):
         features, labels, ids = next(it)
         features = features.to(args.device)
         labels = labels.numpy()
         ids = ids.numpy()
-        with torch.no_grad():
+        with torch.no_grad() and auto_cast():
```
Comment (NetZissou) [inline src/evaluation/fishnet.py]:
```suggestion
        with torch.no_grad(), auto_cast():
```

File: src/evaluation/herbarium19.py [hunk 3]
```diff
@@ -250,8 +254,12 @@ def get_features(
         images, labels, ids, masks = next(it)
         images = images.to(args.device)
 
-        with torch.no_grad():
+        if cast_dtype is not None:
+            images = images.to(dtype=cast_dtype)
+
+        with torch.no_grad() and autocast():
```
Comment (NetZissou) [inline src/evaluation/herbarium19.py]:
```suggestion
        with torch.no_grad(), autocast():
```

**Review by NetZissou — CHANGES_REQUESTED** [review]
A bug: `with torch.no_grad() and auto_cast()` doesn't work as intended

Python `and` operator does not combine the two torch context managers. It evaluates them as boolean expressions, since `torch.no_grad()` is truthy, `and` short-circuits and only `auto_cast()` is entered. `no_grad()` is never activated, so gradients remain enabled during inference. 

@vimar-gu you [mentioned earlier](https://github.com/Imageomics/TreeOfLife-dev/issues/99#issuecomment-4017729326) that the classification got slower, I wonder if this was the bug that caused it. 

```python
# Verified with torch 2.6:

>>> with torch.no_grad() and auto_cast():
        print(torch.is_grad_enabled())
True   # bug: grad still enabled

>>> with torch.no_grad(), auto_cast():
        print(torch.is_grad_enabled())
False  # correct!
```

The fix is basically replace `and` with comma `,` to enable multiple context managers 
```python
# This:
with A(), B():                                                                     
    body        

# Is equivalent to:
with A():
    with B():                                                                      
        body     
```

**Comment by vimar-gu** [comment]
Ha! Codex suggested the `and` operation :( (or Copilot. I forgot)
But the slower speed was not tested on these tasks. It happened during few-shot species classification after the embeddings were extracted. So I guess that was caused by frequent half-precision calculations.

**Review by NetZissou — APPROVED** [review]
LGTM!
